### PR TITLE
Add ability for flow generics to break

### PR DESCRIPTION
**Pull request by vjeux**
While looking at an instability on the React codebase ( https://github.com/sebmarkbage/react/commit/ba1299acc2c3b66ec1e093c201bcede421a8caa0#diff-2550aa3d377452ae29361f5e53c51c10 ), I realized that we don't let them break which makes the code look weird.

So I added the ability for them to break :)

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1737,8 +1737,28 @@ function genericPrintNoParens(path, options, print) {
         ")"
       ]);
     case "TypeParameterDeclaration":
-    case "TypeParameterInstantiation":
-      return concat(["<", join(", ", path.map(print, "params")), ">"]);
+    case "TypeParameterInstantiation": {
+      const shouldInline =
+        n.params.length === 1 &&
+        n.params[0].type === "ObjectTypeAnnotation";
```
Comment (vjeux) [inline src/printer.js]: It looks better when they hug ("deleted" one here)

```diff
 function isActive(
   ad: {
-    state: $Keys<{
-      PAUSED: string,
-      ACTIVE: string,
-      DELETED: string
-    }>
+    state: $Keys<
+      {
+        PAUSED: string,
+        ACTIVE: string,
+        DELETED: string
+      }
+    >
   }
 ): boolean {
   return ad.state === "ACTIVE";
 }
```

File: tests/flow/new_react/fakelib/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -12,7 +12,12 @@ declare var $React: $Exports<\\"react\\">; // fake import
 // Strawman: revised definition of $jsx (alternatively, React.Element).
 // Using bounded poly to specify a constraint on a type parameter, and
 // existentials to elide type arguments.
-type _ReactElement<DefaultProps, Props, Config: $Diff<Props, DefaultProps>, C: $React.Component<DefaultProps, Props, any>> = $React.Element<Config>;
+type _ReactElement<
+  DefaultProps,
+  Props,
+  Config: $Diff<Props, DefaultProps>,
+  C: $React.Component<DefaultProps, Props, any>
+> = $React.Element<Config>;
```
Comment (vjeux) [inline tests/flow/new_react/fakelib/__snapshots__/jsfmt.spec.js.snap]: <3

File: src/printer.js [hunk 3]
```diff
@@ -1559,7 +1559,7 @@ function genericPrintNoParens(path, options, print) {
         );
       }
 
-      return concat(parts);
+      return group(concat(parts));
```
Comment (vjeux) [inline src/printer.js]: By moving the group one level outside, it now first breaks on the arguments and then on the generics. Without this change, it broke a lot of tests with results like this one ("deleted" is the behavior we now have):

```diff
-declare function my_filter<T, P: $Pred<1>>(
-  v: Array<T>,
-  cb: P
-): Array<$Refine<T, P, 1>>;
+declare function my_filter<T, P: $Pred<1>>(v: Array<T>, cb: P): Array<
+  $Refine<T, P, 1>
+>;
```

File: tests/flow_generic/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+    T: (a: A, b: B, c: C, d: D, e: E, f: F) => G // eslint-disable-line space-before-function-paren
+  >(method: T, scope: any, a: A, b: B, c: C, d: D, e: E, f: F): G {}
+};
+"
```
Comment (vjeux) [inline tests/flow_generic/__snapshots__/jsfmt.spec.js.snap]: Used to be printed

```js
var X = {
  perform: function<A, B, C, D, E, F, G, T: (
    a: A,
    b: B,
    c: C,
    d: D,
    e: E,
    f: F
  ) => G>(method: T, scope: any, a: A, b: B, c: C, d: D, e: E, f: F): G {} // eslint-disable-line space-before-function-paren
};
```

**Comment by jlongster** [comment]
Sounds good to me!

**Pull request merged**